### ISSUE TITLE
Type inference problem with `map!`

### ISSUE THREAD

**Issue by dtolnay** [Geal/nom#523]
```rust
#[macro_use]
extern crate nom;

named!(snafu<&str, usize>, map!(tag!("nom"), |s| s.len()));
```

```
error: the type of this value must be known in this context
 --> src/main.rs:4:50
  |
4 | named!(snafu<&str, usize>, map!(tag!("nom"), |s| s.len()));
  |                                                  ^^^^^^^
```

It works with `|s: &str| s.len()` but there should be no reason Rust should need help figuring this out.

Here is our workaround from syn: https://github.com/dtolnay/syn/commit/bc7d7d9877c305808eb6f27fb3da152ef5fc9fb6.

**Comment by Geal** [comment]
I saw the same issue today in another project, but `map!` has not changed for months (maybe years). Maybe something changed in recent nightlies?

**Comment by dtolnay** [comment]
Nope, this was always broken. The approach in https://github.com/dtolnay/syn/commit/bc7d7d9877c305808eb6f27fb3da152ef5fc9fb6 should fix it.

**Comment by Geal** [comment]
this should be fixed with 740287f

**Comment by Geal** [comment]
available in nom 3.1. Thanks for your help!
